codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/rustlang/RustTypes.kt

```diff
@@ -172,6 +172,45 @@ fun RustType.asOptional(): RustType = when (this) {
     else -> RustType.Option(this)
 }
 
+/** Converts type to a reference */
+fun RustType.asRef(): RustType = when (this) {
+    is RustType.Reference -> this
+    is RustType.Option -> RustType.Option(member.asRef())
+    else -> RustType.Reference(null, this)
+}
+
+/** Converts type to its Deref target */
+fun RustType.asDeref(): RustType = when (this) {
```

Comment:
given ref and deref which are so similar, can you add a couple examples to the docs? of how it's different?